File: FrontendRazorPage/Core/Services/QuestionClientService.cs
using System;
using System.Collections.Generic;
using System.Net.Http;
using System.Net.Http.Json;
using System.Threading.Tasks;
using FrontendRazorPage.Core.Services;
using FrontendRazorPage.Models; // Chứa dữ liệu giao diện hoặc DTO thu gọn của bạn
using FrontendRazorPage.Models.AdminModel; // Chứa bảng gốc Models

namespace FrontendRazorPage.Services
{
    public class QuestionClientService : BaseClientService
    {
       

        // 🌟 SỬA ĐỔI 1: Trỏ chính xác về đúng Endpoint Quản trị câu hỏi tại Backend
        private readonly string _apiBaseUrl = "api/QuestionAdmin";

        public QuestionClientService(HttpClient httpClient) : base(httpClient) { }  
        

        // 🔐 1. [ADMIN] Gọi API lấy danh sách câu hỏi theo bài học dạng đầy đủ Model gốc
        public async Task<List<QuestionAdminModel>> GetQuestionsByLessonForAdminAsync(int lessonId)
        {
            if (lessonId <= 0) return new List<QuestionAdminModel>();
            try
            {
                // Gọi đầu api/QuestionAdmin/admin/lesson/{lessonId} công khai hôm trước
                var response = await _httpClient.GetFromJsonAsync<List<QuestionAdminModel>>($"{_apiBaseUrl}/admin/lesson/{lessonId}");
                return response ?? new List<QuestionAdminModel>();
            }
            catch (HttpRequestException)
            {
                return new List<QuestionAdminModel>();
            }
        }

        // 🌐 2. [USER] Gọi API lấy danh sách câu hỏi theo bài học dạng rút gọn DTO
        // 🌟 SỬA ĐỔI 2: Đổi kiểu trả về thành List<QuestionDTO> (hoặc Model giao diện thích hợp của học viên)
        public async Task<List<QuestionModel>> GetQuestionsByLessonForUserAsync(int lessonId)
        {
            if (lessonId <= 0) return new List<QuestionModel>();
            try
            {
                var response = await _httpClient.GetFromJsonAsync<List<QuestionModel>>($"{_apiBaseUrl}/user/lesson/{lessonId}");
                return response ?? new List<QuestionModel>();
            }
            catch (HttpRequestException)
            {
                return new List<QuestionModel>();
            }
        }

        // 🔍 3. [CHUNG] Gọi API lấy chi tiết một câu hỏi theo ID hệ thống
        public async Task<QuestionAdminModel?> GetQuestionByIdAsync(int id)
        {
            try
            {
                return await _httpClient.GetFromJsonAsync<QuestionAdminModel>($"{_apiBaseUrl}/{id}");
            }
            catch (HttpRequestException)
            {
                return null;
            }
        }

        // ➕ 4. [ADMIN] Gửi Request thêm mới toàn bộ cụm dữ liệu câu hỏi lồng kèm đáp án
        public async Task<bool> CreateQuestionAsync(QuestionAdminModel question)
        {
            try
            {
                var response = await _httpClient.PostAsJsonAsync(_apiBaseUrl, question);
                return response.IsSuccessStatusCode;
            }
            catch (HttpRequestException)
            {
                return false;
            }
        }

        // ✏️ 5. [ADMIN] Gửi Request cập nhật sửa đổi, đồng bộ mảng câu trả lời Answers
        public async Task<bool> UpdateQuestionAsync(int id, QuestionAdminModel question)
        {
            try
            {
                var response = await _httpClient.PutAsJsonAsync($"{_apiBaseUrl}/{id}", question);
                return response.IsSuccessStatusCode;
            }
            catch (HttpRequestException)
            {
                return false;
            }
        }

        // ❌ 6. [ADMIN] Gửi lệnh xóa câu hỏi vĩnh viễn khỏi hệ thống database
        public async Task<bool> DeleteQuestionAsync(int id)
        {
            try
            {
                var response = await _httpClient.DeleteAsync($"{_apiBaseUrl}/{id}");
                return response.IsSuccessStatusCode;
            }
            catch (HttpRequestException)
            {
                return false;
            }
        }
    }
}

File: FrontendAngular/src/app/Core/Services/question-client-service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, of } from 'rxjs'; // ◄ Dùng Observable và of từ RxJS
import { catchError, map } from 'rxjs/operators'; // ◄ Các toán tử bắt lỗi và biến đổi dữ liệu
import { BaseService } from './BaseService';
import { QuestionModel } from '../../Models/question-model';
import { QuestionAdminModel } from '../../Models/AdminModel/question-admin-model';

@Injectable({
    providedIn: 'root'
})
export class QuestionClientService extends BaseService {

    // 🌟 Định nghĩa chuẩn Endpoint gốc cho controller QuestionAdmins
    private readonly _apiBase = `${this._apiBaseUrl}/QuestionAdmin`;

    constructor(private _httpClient: HttpClient) {
        super(); // Gọi constructor của lớp cha BaseService
    }

    /**
     * 🔐 1. [ADMIN] Gọi API lấy danh sách câu hỏi theo bài học dạng đầy đủ Model gốc
     */
    public getQuestionsByLessonForAdminAsync(lessonId: number): Observable<QuestionAdminModel[]> {
        if (lessonId <= 0) return of([]);

        return this._httpClient.get<QuestionAdminModel[]>(`${this._apiBase}/admin/lesson/${lessonId}`).pipe(
            catchError(() => of([])) // Nếu lỗi, trả về mảng rỗng
        );
    }

    /**
     * 🌐 2. [USER] Gọi API lấy danh sách câu hỏi theo bài học dạng rút gọn
     */
    public getQuestionsByLessonForUserAsync(lessonId: number): Observable<QuestionModel[]> {
        if (lessonId <= 0) return of([]);

        return this._httpClient.get<QuestionModel[]>(`${this._apiBase}/user/lesson/${lessonId}`).pipe(
            catchError(() => of([]))
        );
    }

    /**
     * 🔍 3. [CHUNG] Gọi API lấy chi tiết một câu hỏi theo ID hệ thống
     */
    public getQuestionByIdAsync(id: number): Observable<QuestionAdminModel | null> {
        return this._httpClient.get<QuestionAdminModel>(`${this._apiBase}/${id}`).pipe(
            catchError(() => of(null)) // Nếu lỗi, trả về null để tránh sập app
        );
    }

    /**
     * ➕ 4. [ADMIN] Gửi Request thêm mới toàn bộ cụm dữ liệu câu hỏi lồng kèm đáp án
     */
    public createQuestionAsync(question: QuestionAdminModel): Observable<boolean> {
        return this._httpClient.post(`${this._apiBase}`, question, { observe: 'response' }).pipe(
            map(response => response.ok), // Nếu status code dạng 2xx (Thành công) -> trả về true
            catchError(() => of(false))   // Nếu lỗi HTTP -> trả về false
        );
    }

    /**
     * ✏️ 5. [ADMIN] Gửi Request cập nhật sửa đổi, đồng bộ mảng câu trả lời Answers
     */
    public updateQuestionAsync(id: number, question: QuestionAdminModel): Observable<boolean> {
        return this._httpClient.put(`${this._apiBase}/${id}`, question, { observe: 'response' }).pipe(
            map(response => response.ok),
            catchError(() => of(false))
        );
    }

    /**
     * ❌ 6. [ADMIN] Gửi lệnh xóa câu hỏi vĩnh viễn khỏi hệ thống database
     */
    public deleteQuestionAsync(id: number): Observable<boolean> {
        return this._httpClient.delete(`${this._apiBase}/${id}`, { observe: 'response' }).pipe(
            map(response => response.ok),
            catchError(() => of(false))
        );
    }

    /**
     * 🌾 7. Bắn URL dạng tham số query: /questions?grammarId=1&questionType=1
     */
    public getQuestionsByGrammarAsync(grammarId: number, questionType: number = 0): Observable<QuestionModel[]> {
        if (grammarId <= 0) return of([]);

        const url = `${this._apiBase}/questions?grammarId=${grammarId}&questionType=${questionType}`;
        return this._httpClient.get<QuestionModel[]>(url).pipe(
            catchError(() => of([]))
        );
    }
}